pkg/paths/paths.go

```diff
@@ -0,0 +1,111 @@
+// Copyright (C) 2018 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package paths
+
+import (
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/hmac"
+	"crypto/sha512"
+	"encoding/base64"
+)
+
+// Encrypt the given path with the given key
+func Encrypt(path []string, key []byte) (encryptedPath []string, err error) {
+	encryptedPath = make([]string, len(path))
+	secret := key
+	for i, seg := range path {
+		encryptedPath[i], err = encrypt(seg, secret)
+		if err != nil {
+			return nil, err
+		}
+		secret = deriveSecret(secret, seg)
+	}
+	return encryptedPath, nil
+}
+
+// Decrypt the given encrypted path with the given key
+func Decrypt(encryptedPath []string, key []byte) (path []string, err error) {
+	path = make([]string, len(encryptedPath))
+	secret := key
+	for i, seg := range encryptedPath {
+		path[i], err = decrypt(seg, secret)
+		if err != nil {
+			return nil, err
+		}
+		secret = deriveSecret(secret, path[i])
+	}
+	return path, nil
+}
+
+// DeriveKey derives the key for the given path from the given root key
+func DeriveKey(key []byte, path []string) (derivedKey []byte) {
+	derivedKey = key
+	for _, seg := range path {
+		derivedKey = deriveSecret(derivedKey, seg)
+	}
+	return derivedKey
+}
+
+func encrypt(text string, secret []byte) (cipherText string, err error) {
+	key, nonce := getAESGCMKeyAndNonce(secret)
+	block, err := aes.NewCipher(key)
+	if err != nil {
+		return "", err
+	}
+	aesgcm, err := cipher.NewGCM(block)
+	if err != nil {
+		return "", err
+	}
+	data := aesgcm.Seal(nil, nonce, []byte(text), nil)
+	cipherText = base64.RawURLEncoding.EncodeToString(data)
+	// prepend version number to the cipher text
+	cipherText = "1" + cipherText
+	return cipherText, nil
+}
+
+func decrypt(cipherText string, secret []byte) (text string, err error) {
+	if cipherText == "" {
+		return "", Error.New("empty cipher text")
+	}
+	// check the version number, only "1" is supported for now
+	if cipherText[0] != '1' {
+		return "", Error.New("invalid version number")
+	}
+	data, err := base64.RawURLEncoding.DecodeString(cipherText[1:])
+	if err != nil {
+		return "", err
+	}
+	key, nonce := getAESGCMKeyAndNonce(secret)
+	block, err := aes.NewCipher(key)
+	if err != nil {
+		return "", err
```

Comment:
with the errors, should we do Error.Wrap() ?

Reply:
Yes, I guess.